File: LeetCode/50.Pow(x,_n)/50.Pow(x,_n).swift

// iterative
class Solution {
    func myPow(_ x: Double, _ n: Int) -> Double {
        if n == 0 { return 1.0 }
        if x == 0 { return 0.0 }

        var e: Int = n, base: Double = x
        if e < 0 {
            e = -e
            base = 1.0 / base
        }

        var res = 1.0
        while e > 0 {
            if e & 1 == 1 {
                res *= base
            }

            base *= base
            e >>= 1
        }

        return res
    }
}

// recursive
class Solution2 {
    func myPow(_ x: Double, _ n: Int) -> Double {
        if n == 0 { return 1.0 }
        if n < 0 { return 1.0 / myPow(x, -n) }

        if n & 1 == 1 {
            return x * myPow(x * x, (n - 1) / 2)
        } else {
            return myPow(x * x, n / 2)
        }
    }
}

File: LeetCode/50.Pow(x,_n)/50.Pow(x,_n).py
#!/usr/bin/env python3


class Solution:
    def myPow(self, x: float, n: int) -> float:

        def pos_pow(x, n):
            # print(f"n:{n}")
            if n == 0:
                return 1.0
            half = pos_pow(x, n // 2)
            if n % 2 == 0:
                return half * half
            else:
                return half * half * x

        if n < 0:
            return 1 / pos_pow(x, -n)
        else:
            return pos_pow(x, n)


class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n == 0:
            return 1.0
        elif n < 0:
            x = 1 / x
            n = -n

        res = 1.0
        while n > 0:
            if n % 2 == 1:
                res *= x
            x *= x
            n //= 2
            # print(f"x:{x}")
        return res
